applications/ContactStructuralMechanicsApplication/custom_processes/aalm_adapt_penalty_value_process.cpp

```diff
@@ -25,56 +25,55 @@ void AALMAdaptPenaltyValueProcess::Execute()
-    for(int i = 0; i < static_cast<int>(nodes_array.size()); ++i) 
-    {
+    for(int i = 0; i < static_cast<int>(nodes_array.size()); ++i) {
         auto it_node = nodes_array.begin() + i;
         
         // Initial value
         const double penalty_parameter = initialize ? initial_penalty_parameter : it_node->GetValue(INITIAL_PENALTY);
         
         // Weighted values
         const double nodal_area = it_node->GetValue(NODAL_AREA);
-        const double current_gap = it_node->FastGetSolutionStepValue(WEIGHTED_GAP)/nodal_area;
-        const double previous_gap = it_node->FastGetSolutionStepValue(WEIGHTED_GAP, 1)/nodal_area;
-        
-        // Nodal H x gap factor
-        const double max_gap = max_gap_factor * it_node->FastGetSolutionStepValue(NODAL_H); // NOTE: This value must be studied
-        
-//         // DEBUG
-//         std::cout << "current_gap: " << current_gap << "\tprevious_gap: " << previous_gap << "\tmax_gap: " << max_gap << std::endl;
-//         std::cout << "(current_gap * previous_gap): " << (current_gap * previous_gap) << std::endl;
-//         std::cout << "std::abs(previous_gap > max_gap): " << std::abs(previous_gap > max_gap) << std::endl;
-//         std::cout << "std::abs(current_gap > max_gap): " << std::abs(current_gap > max_gap) << std::endl;
-        
-        if ((current_gap * previous_gap) < 0.0) 
-        {
-            if (std::abs(previous_gap) > max_gap) // NOTE: The abs is deduced from the paper (the algorithm is without abs)
-                it_node->SetValue(INITIAL_PENALTY, std::abs(penalty_parameter * previous_gap / (current_gap) * (std::abs(current_gap) + max_gap)/(current_gap - previous_gap)));
-            else
-                it_node->SetValue(INITIAL_PENALTY, std::abs(penalty_parameter * previous_gap / (10.0 * current_gap)));
-        }
-        else if (std::abs(current_gap) > max_gap) // NOTE: The abs is deduced from the paper (the algorithm is without abs)
-        {
-            if (std::abs(current_gap - previous_gap) > std::max(current_gap/10.0, std::max(previous_gap/1.0, 5 * max_gap)))
-                it_node->SetValue(INITIAL_PENALTY, 2.0 * penalty_parameter);
-            else if ((std::abs(current_gap) <= std::abs(previous_gap) * 1.01 || std::abs(current_gap) >= std::abs(previous_gap) * 0.99) && (std::abs(current_gap) < 10.0 *  max_gap))
-                it_node->SetValue(INITIAL_PENALTY, penalty_parameter * std::pow((std::sqrt(std::abs(current_gap)/max_gap - 1.0) + 1.0), 2.0));
-            else if (std::abs(current_gap) > std::abs(previous_gap) * 1.01)
-                it_node->SetValue(INITIAL_PENALTY, 2.0 * penalty_parameter * (previous_gap/current_gap));
-            else
-                it_node->SetValue(INITIAL_PENALTY, penalty_parameter * (std::sqrt(std::abs(current_gap)/max_gap - 1.0) + 1.0));
+        if (nodal_area > std::numeric_limits<double>::epsilon()) {
+            const double current_gap = it_node->FastGetSolutionStepValue(WEIGHTED_GAP)/nodal_area;
```

Comment:
are u sure you should not obtain WEIGHTED_GAP by solving the consistent mass matrix? what happens if elements are only partially active

Reply:
Technically that is already taken into account during the computation of the WEIGHTED_GAP, that is done using ExplicitContribution

Reply:
then why are u dividing again by the nodal area?

Reply:
Because the weighted_gap is not a real gap, is ponderated on the integration, in order to have a physical meaning (meters and so on) it is necessary to divide by the "NODAL_AREA", which is the proportion of the lumped mass matrix during the dual LM solution